docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+
+The PE approach reduces the original right-hand side of a transparent function via term rewriting. The right hand side is conceptually the term as it is written down, closed over the environment where the transparent function was defined. This is implemented by rewriting and then re-type-checking the original untyped tree, but with typed leaves that refer to the tree's free variables. The re-type-checking with more specific argument types can lead to a type derivation that is quite different
+than the original typing of the transparent method's right hand side. Types might be inferred to be more specific, overloading resolution could pick different alternatives, and implicit searches might yield different results or might be elided altogether.
+This flexibility is what makes code specialization work smoothly.
+
+By constrast, the DT approach typechecks the right hand side of a transparent function in a special way. Instead of
+computing types as usual, it "lifts" every term into a type that mirrors it. To do this, it needs to create type equivalents
+of all relevant term forms including applications, conditionals, match expressions, and the various forms of pattern matches.
+The end result of the DT approach is then a type `{t}` that represents essentially the right-hand side term `t` itself, or at least all portions that might contribute to that type.  The type has to be constructed in such a way that the result type of every application
+of a transparent function can be read off from it.
+An application of a transparent function then needs to just instantiate that type, whereas the term is not affected at all, and the function is called as usual. This means there is a guarantee that the
+semantics of a transparent call is the same as a normal call, only the type is better. On the other hand, one cannot play
+tricks such as code specialization, since there is nothing to specialize.
+Also, implicit matches would not fit into this scheme, as they require term rewritings. Another concern is how much additional complexity would be caused by mirroring the necessary term forms in types and tweaking type inference to work with the new types.
+
+To summarize, here are some advantages of DT over PE:
+
+ + It avoids code explosion by design (can be solved in PE but requires extra work).
+ + Expansion only works on types, so this might well be faster at compile-time than PE's term rewriting.
+ + It gives a guarantee that `transparent` is semantics preserving.
+ + The typeof `{t}` operator is interesting in its own right.
+
+By contrast, here are some advantages of PE over DT:
+
+ + It uses the standard type checker and typing rules with no need to go to full dependent types
```

Comment:
I think it's a but unfair to claim "the use of standard type checker and typing rules" for PE. Indeed, PE wouldn't perform any (type) specialization without the Rewrite Rules which are neither part of the standard type checker nor of the typing rules. Also, these rewrites rules have a 1-1 correspondence to new typing rules we would need for "full term-dependencies".